moveit_ros/planning_interface/test/python_move_group_planning.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python
+
+import unittest
+import numpy as np
+import rospy
+import rostest
+import os
+
+from moveit_msgs.msg import MoveItErrorCodes
+import moveit_commander
+
+from geometry_msgs.msg import PoseStamped
+
+
+class PythonMoveGroupPlanningTest(unittest.TestCase):
+    @classmethod
+    def setUpClass(self):
+        PLANNING_GROUP = "panda_arm"
+        self.group = moveit_commander.MoveGroupCommander(PLANNING_GROUP)
+        self.planning_scene_interface = moveit_commander.PlanningSceneInterface(
+            synchronous=True
+        )
+
+    @classmethod
+    def tearDown(self):
+        pass
+
+    def test_planning_with_collision_objects(self):
+        # Add obstacle to the world
+        ps = PoseStamped()
+        ps.header.frame_id = "world"
+        ps.pose.position.x = 0.4
+        ps.pose.position.y = 0.1
+        ps.pose.position.z = 0.25
+        self.planning_scene_interface.add_box(
+            name="box1", pose=ps, size=(0.1, 0.1, 0.5)
+        )
+
+        # Attach object to robot's TCP
+        ps2 = PoseStamped()
+        tcp_link = self.group.get_end_effector_link()
+        ps2.header.frame_id = tcp_link
+        ps2.pose.position.z = 0.15
+        self.planning_scene_interface.attach_box(
+            link=tcp_link,
+            name="box2",
+            pose=ps2,
+            size=(0.1, 0.1, 0.1),
+            touch_links=["panda_rightfinger", "panda_leftfinger"],
+        )
+
+        # Plan a motion where the attached object 'box2' collides with the obstacle 'box1'
+        target_pose = self.group.get_current_pose(tcp_link)
+        target_pose.pose.position.y += 0.1
+
+        # Set planner to be Pilz's Linear Planner
+        self.group.set_planning_pipeline_id("pilz_industrial_motion_planner")
+        self.group.set_planner_id("LIN")
```

Comment:
I would prefer to keep this test limited to `PTP` unless the code paths for checking collisions are strictly separated between the two planners. `LIN` relies on point-wise IK requests that add an unnecessary complexity here and might make the test flaky for unrelated reasons.

Reply:
I guess, in a sense they are separated. `PTP` only checks collision of the target pose if the target pose is a Cartesian pose, otherwise it does not check collisions at all. I think `LIN` checks collisions for intermediate points. 
Now, neither checks collision of attached objects, as the only collision checks done in the Pilz planners are `checkSelfCollision`. So, after all, just testing `PTP` is fine I suppose.

Reply:
>  PTP only checks collision of the target pose if the target pose is a Cartesian pose, otherwise it does not check collisions at all. I think LIN checks collisions for intermediate points.
Now, neither checks collision of attached objects, as the only collision checks done in the Pilz planners are checkSelfCollision

Wait, what? I was only peripherally involved with #2824, but I thought that's exactly what it was meant to solve?
Also the test case you propose here relies on world collision checking. 

Does all of this still rely on MoveIt's post-planning check of the entire trajectory? :open_mouth:

Reply:
No, right now, it does rely only on MoveIt's post-planning check. The fix in #2824 was to make the post-planning work properly. 

Now, if that post-planning is only optional, then I guess it is okay to add the same validation of the entire trajectory **here**, where the internal planner has returned a solution but before returning to the planning_pipeline 

https://github.com/ros-planning/moveit/blob/8c499f5d3dea8a75c6219be765d5a824e34d9056/moveit_planners/pilz_industrial_motion_planner/src/trajectory_generator.cpp#L305-L317

It would work for every type of planner LIN, PTP, and CIRC.

I think this approach is acceptable as Pilz planners basically assume the existence of one possible solution, so checking that solution before returning should be safe and enough.

Reply:
Without looking at the code, the idea was, that IK solvers use the collision-check-callback to choose a suitable solution that is closest to the seed (previous point). It should take into account attached objects, of course (otherwise it should be fixed). In contrast PTP only does that check once for the goal, if it is given in cartesian space.
Relying on the pipeline for collision checking was mainly a decision to not duplicate computation effort and speed-up planning, if I remember correctly.

Reply:
makes sense. let's keep it that way. If users disable the optional post-planning check, they should be aware of the nature of results from their planners.